### PR TITLE
Simplify output capturing

### PR DESCRIPTION
**Pull request by m-ou-se**
This is a sequence of incremental improvements to the unstable/internal `set_panic` and `set_print` mechanism used by the `test` crate:

1. Remove the `LocalOutput` trait and use `Arc<Mutex<dyn Write>>` instead of `Box<dyn LocalOutput>`. In practice, all implementations of `LocalOutput` were just `Arc<Mutex<..>>`. This simplifies some logic and removes all custom `Sink` implementations such as `library/test/src/helpers/sink.rs`. Also removes a layer of indirection, as the outermost `Box` is now gone. It also means that locking now happens per `write_fmt`, not per individual `write` within. (So `"{} {}\n"` now results in one `lock()`, not four or more.)

2. Since in all cases the `dyn Write`s were just `Vec<u8>`s, replace the type with `Arc<Mutex<Vec<u8>>>`. This simplifies things more, as error handling and flushing can be removed now. This also removes the hack needed in the default panic handler to make this work with `::realstd`, as (unlike `Write`) `Vec<u8>` is from `alloc`, not `std`.

3. Replace the `RefCell`s by regular `Cell`s. The `RefCell`s were mostly used as `mem::replace(&mut *cell.borrow_mut(), something)`, which is just `Cell::replace`. This removes an unecessary bookkeeping and makes the code a bit easier to read.

4. Merge `set_panic` and `set_print` into a single `set_output_capture`. Neither the test crate nor rustc (the only users of this feature) have a use for using these separately. Merging them simplifies things even more. This uses a new function name and feature name, to make it clearer this is internal and not supposed to be used by other crates.

Might be easier to review per commit.

### CODE REVIEW

File: library/std/src/io/stdio.rs [hunk 1]
```diff
@@ -988,36 +909,26 @@ pub(crate) fn clone_io() -> (Option<Box<dyn LocalOutput>>, Option<Box<dyn LocalO
...
 ///
 /// However, if the actual I/O causes an error, this function does panic.
-fn print_to<T>(
-    args: fmt::Arguments<'_>,
-    local_s: &'static LocalKey<RefCell<Option<Box<dyn LocalOutput>>>>,
-    global_s: fn() -> T,
-    label: &str,
-) where
+fn print_to<T>(args: fmt::Arguments<'_>, global_s: fn() -> T, label: &str)
+where
```
Comment (o752d) [inline library/std/src/io/stdio.rs:913]:
```suggestion
fn print_to<T>(
  args: fmt::Arguments<'_>,
  global_s: fn() -> T,
  label: &str
) where
```
Is there specific guidance on where to nest related elements for readability?

Comment (m-ou-se) [reply library/std/src/io/stdio.rs:913]: > Is there specific guidance on where to nest related elements for readability?

Yes, we simply do whatever `rustfmt` does. (Run by `./x.py fmt` in the `rust` repository.)

Your suggestion would not pass the automated style check. It's automatically checked that the formatting matches what `rustfmt` would do.

Comment (o752d) [reply library/std/src/io/stdio.rs:913]: Ok thanks!

**Comment by KodrAus** [comment]
> Merge set_panic and set_print into a single set_output_capture

I think these are two unstable APIs that are actually relied upon out in the wild. We have reserved the right to tinker with them, but I'd be interested to see how widely they're actually used! I'll kick off a crater run to see what impact this has.

@craterbot check

**Comment by m-ou-se** [comment]
> I think these are two unstable APIs that are actually relied upon out in the wild.

Those APIs were already broken recently by https://github.com/rust-lang/rust/pull/75172 / https://github.com/rust-lang/rust/pull/78227

**Comment by m-ou-se** [comment]
Starting crater with `start=beta` instead of the parent commit of this PR, because the function signatures were already changed/broken two weeks ago. (Hopefully that won't report too many unrelated changes.)

@craterbot check start=beta

**Comment by m-ou-se** [comment]
@craterbot check start=beta end=try#3e22d1f9053c0fc87629b3c6d5d9567856368e5d

**Comment by m-ou-se** [comment]
That crater run resulted in quite a few irrelevant failures, because it compaired against beta. These are the only four that had errors in their logs about missing `set_panic` or `set_print`:

- [`wasm_glue`](https://docs.rs/wasm-glue/0.1.0/wasm_glue/):
   This crate's main purpose is to use `set_print` and `set_panic` to work around some issues with stdout/stderr on wasm. This crate doesn't seem to be very popular though.
- `cis198-2016s/demo-xcompile-android`:
   Fails because it depends on an old version of [the deprecated `android-rs-glue` project](https://github.com/rust-windowing/android-rs-glue) that used `set_print`/`set_panic`. Newer versions and the project that superseded it don't use those functions.
- `reitermarkus/lndir` and `rvolz/rustrt`:
   These both fail because of their dependency on an old version of `cucumber_rust`. Versions since may 2019 [don't use `set_print` or `set_panic` anymore](https://github.com/bbqsrc/cucumber-rust/commit/aa4bba37f9857cca99dd69cfcaa357e6adf044e3).

**Comment by Mark-Simulacrum** [comment]
FWIW we have access to by-commit artifacts so you can get a smaller commit range, but ultimately it probably wouldn't help too much. If we want a better run I would probably revert the breakage in a pr, try commit that, then compare against that in crater.

Let me know if you need more help here.

**Comment by m-ou-se** [comment]
In this case I knew what the error would be, so it was as simple as just downloading the build-fail tar.gz from crater and grepping for `set_panic` and `set_print`. Only four crates matched.

**Comment by KodrAus** [comment]
That's a lot fewer than I was expecting! Thanks for looking into this @m-ou-se

**Comment by KodrAus** [comment]
Ok, this looks good to me. I was worried about churning an unstable hack that I thought was pretty widely relied on, but it turns out I was wrong!

@bors r+

**Pull request merged**